src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotPolicyStep.java

```diff
@@ -0,0 +1,45 @@
+package bio.terra.service.snapshot.flight.delete;
+
+import bio.terra.common.exception.FeatureNotImplementedException;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.dataset.flight.DatasetWorkingMapKeys;
+import bio.terra.service.job.DefaultUndoStep;
+import bio.terra.service.policy.PolicyService;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DeleteSnapshotPolicyStep extends DefaultUndoStep {
+  private static final Logger logger = LoggerFactory.getLogger(DeleteSnapshotPolicyStep.class);
+  private final DatasetService datasetService;
+  private final PolicyService policyService;
+  private final UUID snapshotId;
+
+  public DeleteSnapshotPolicyStep(
+      DatasetService datasetService, PolicyService policyService, UUID snapshotId) {
+    this.datasetService = datasetService;
+    this.policyService = policyService;
+    this.snapshotId = snapshotId;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext)
+      throws InterruptedException, RetryException {
+    FlightMap map = flightContext.getWorkingMap();
+    UUID datasetId = map.get(DatasetWorkingMapKeys.DATASET_ID, UUID.class);
+    Dataset dataset = datasetService.retrieve(datasetId);
+    if (dataset.isSecureMonitoringEnabled()) {
+      try {
+        policyService.deletePao(snapshotId);
+      } catch (FeatureNotImplementedException ex) {
```

Comment:
Ah! Seeing this `FeatureNotImplementedException` handling made me think…

I know that we call `policyService.deletePdao` in the undo method for `CreateSnapshotPolicyStep`.  Should we have that handling there as well, so that if a snapshot creation flight gets undone it doesn't fail dismally?

Reply:
Oh you're totally right! I'll add that in

Reply:
Thanks for adding that in!

Now that I'm seeing the code for both `CreateSnapshotPolicyStep` and `DeleteSnapshotPolicyStep` I am struck at how similar they are.  I could be missing something, but it looks like one step's do is the other's undo, and vice-versa.  What do you think about deriving both steps from a common class to facilitate code sharing and ensure common error handling?

A higher-level musing on errors -- I am struggling to see the value of the `TpsService` methods throwing `FeatureNotImplementedException`s if we're going to catch them everywhere.  It seems like it makes for more code clean-up down the line when we remove the feature flag, where instead we could maybe exit early from those methods with a log but not throw.  Happy to be convinced otherwise though!

Reply:
I like your suggestion to consolidate the code! Especially since we're planning to add more logic into these steps for other policy types so I could see this getting out of sync by mistake.

For the `FeatureNotImplementedException`s, I missed a spot where I should have caught that exception so I think that proves your point 😅 I was thinking it would result in a lot of repetition in the TpsService methods, but we're planning to remove the logic anyways and it will all be in the same class.

Reply:
Awesome! It's nice to be able to take advantage of the symmetry between creating and deleting snapshots here.